src/epic/distances/divergence_free.py

```diff
@@ -0,0 +1,217 @@
+        network) to minimize the L2 norm of the shaped reward function.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, max(self.state_dim * 4, 128)),
+                nn.GELU(),
+                Residual(nn.Linear(max(self.state_dim * 4, 128), max(self.state_dim * 4, 128))),
+                nn.GELU(),
+                nn.Linear(max(self.state_dim * 4, 128), 1),
+            ],
+        )
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        net.to(device)
+
+        max_epochs = 10000
+        optimizer = optim.AdamW(net.parameters(), lr=6e-4)
+        scheduler = optim.lr_scheduler.LambdaLR(
+            optimizer, lambda epoch: 0.5 if (epoch > 5000 and epoch < 7500) else 0.25 if (epoch > 7500) else 1.0
+        )
+
+        state_sample, action_sample, next_state_sample, done_sample = self.coverage_sampler.sample(
+            n_samples_can,
+        )
+
+        state_sample_tensor = utils.float_tensor_from_numpy(state_sample, device)
+        next_state_sample_tensor = utils.float_tensor_from_numpy(next_state_sample, device)
+
+        losses = []
+
+        with torch.inference_mode():
+            if state_sample_tensor.ndim == 1:
+                assert next_state_sample_tensor.ndim == 1
+                state_sample_tensor.unsqueeze_(-1)
+                next_state_sample_tensor.unsqueeze_(-1)
+
+            potential = (self.discount_factor * net(next_state_sample_tensor) - net(state_sample_tensor)).squeeze(-1)
+            pre_training_l2_loss = torch.mean(
+                (
+                    utils.float_tensor_from_numpy(
+                        rew_fn(state_sample, action_sample, next_state_sample, done_sample),
+                        device,
+                    )
+                    + potential
+                )
+                ** 2,
+            )
+            losses.append(pre_training_l2_loss.item())
+
+        for _ in range(max_epochs):
+            potential = (self.discount_factor * net(next_state_sample_tensor) - net(state_sample_tensor)).squeeze(-1)
+            l2_loss = torch.mean(
+                (
+                    utils.float_tensor_from_numpy(
+                        rew_fn(state_sample, action_sample, next_state_sample, done_sample),
+                        device,
+                    )
+                    + potential
+                )
+                ** 2,
+            )
+            l2_loss.backward()
+            optimizer.step()
+            optimizer.zero_grad()
+            losses.append(l2_loss.item())
+
+            # Early stopping if loss has stopped fluctuating
+            if len(losses) >= 500:
+                last_five_losses = losses[-500:]
```

Comment:
These are the last 500 losses, not 5, right? (not sure which of the two you wanted)

Reply:
Yeah this should probably just be renamed (and perhaps the early_stopping window should be configurable).